src/ast_editor.ml

```diff
@@ -0,0 +1,487 @@
+    ~(webview : WebView.t) =
+  let changed_document = TextDocumentChangeEvent.document event in
+  if document_eq document changed_document then
+    transform_to_ast ~document ~webview
+
+let onDidReceiveMessage_listener instance msg ~(document : TextDocument.t) =
+  let ast_editor_state = Extension_instance.ast_editor_state instance in
+  let int_prop name =
+    if Ojs.has_property msg name then
+      Some (Int.of_string (Ojs.string_of_js (Ojs.get_prop_ascii msg name)))
+    else
+      None
+  in
+  match int_prop "selectedOutput" with
+  | Some i -> Ast_editor_state.set_original_mode ast_editor_state (i = 0)
+  | None -> (
+    match Option.both (int_prop "begin") (int_prop "end") with
+    | None -> ()
+    | Some (cbegin, cend) ->
+      let apply_selection editor cbegin cend =
+        let document = TextEditor.document editor in
+        let anchor = Vscode.TextDocument.positionAt document ~offset:cbegin in
+        let active = Vscode.TextDocument.positionAt document ~offset:cend in
+        TextEditor.set_selection editor
+          (Selection.makePositions ~anchor ~active);
+        TextEditor.revealRange editor
+          ~range:(Range.makePositions ~start:anchor ~end_:active)
+          ()
+      in
+      Vscode.Window.visibleTextEditors ()
+      |> List.iter ~f:(fun editor ->
+             let visible_doc = TextEditor.document editor in
+             if (* !original_mode && *) document_eq document visible_doc then
+               apply_selection editor cbegin cend
+             else if
+               (* (not !original_mode) && *)
+               (Ast_editor_state.entry_exists ast_editor_state
+                  ~origin_doc:(doc_string_uri ~document))
+                 ~pp_doc:(doc_string_uri ~document:visible_doc)
+               && not (Ast_editor_state.get_original_mode ast_editor_state)
+             then
+               match Option.both (int_prop "r_begin") (int_prop "r_end") with
+               | None -> ()
+               | Some (rcbegin, rcend) -> apply_selection editor rcbegin rcend))
+
+let on_hover custom_doc webview =
+  let provider =
+    let provideHover ~(document : TextDocument.t) ~(position : Position.t)
+        ~token:_ =
+      let offset = TextDocument.offsetAt document ~position in
+      if document_eq custom_doc document then
+        send_msg "focus" (Ojs.int_to_js offset) ~webview;
+      let hover =
+        Hover.make
+          ~contents:(`MarkdownString (MarkdownString.make ~value:"" ()))
+      in
+      `Value (Some hover)
+    in
+    HoverProvider.create ~provideHover
+  in
+  Vscode.Languages.registerHoverProvider ~selector:(`String "ocaml") ~provider
+
+let activate_hover_mode instance ~document =
+  let ast_editor_state = Extension_instance.ast_editor_state instance in
+  match Ast_editor_state.find_webview_by_doc ast_editor_state ~document with
+  | Some webview -> on_hover document webview
+  | None -> failwith "Webview wasn't found while switching hover mode"
```

Comment:
Under what circumstances is this error possible?

Reply:
It happends when user tries to activate hover mode without opening the AST explorer for instance. 
I added an error message to clarify this 69aae11657ecddc8973ccd56ab8dfcf1a307fe4f .